dbms/src/Functions/FunctionsRegexp.h

```diff
@@ -1512,6 +1532,348 @@ class FunctionStringRegexpInstr : public FunctionStringRegexpBase
+    void setCollator(const TiDB::TiDBCollatorPtr & collator_) override { collator = collator_; }
+    bool useDefaultImplementationForNulls() const override { return false; }
+    size_t getNumberOfArguments() const override { return 0; }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+    {
+        size_t arg_num = arguments.size();
+        if (arg_num < REGEXP_MIN_PARAM_NUM)
+            throw Exception("Too few arguments", ErrorCodes::TOO_LESS_ARGUMENTS_FOR_FUNCTION);
+        else if (arg_num > REGEXP_SUBSTR_MAX_PARAM_NUM)
+            throw Exception("Too many arguments", ErrorCodes::TOO_MANY_ARGUMENTS_FOR_FUNCTION);
+
+        bool has_nullable_col = false;
+        bool has_data_type_nothing = false;
+        bool is_str_arg;
+
+        // Check type of arguments
+        for (size_t i = 0; i < arg_num; ++i)
+        {
+            // Index at 0, 1 and 4 arguments should be string type, otherwise int type.
+            is_str_arg = (i <= 1 || i == 4);
+            checkInputArg(arguments[i], is_str_arg, &has_nullable_col, &has_data_type_nothing);
+        }
+
+        if (has_data_type_nothing)
+            return std::make_shared<DataTypeNullable>(std::make_shared<DataTypeNothing>());
+
+        return std::make_shared<DataTypeNullable>(std::make_shared<DataTypeString>());
+    }
+
+    template <typename ExprT, typename PatT, typename PosT, typename OccurT, typename MatchTypeT>
+    void REGEXP_CLASS_MEM_FUNC_IMPL_NAME(ColumnWithTypeAndName & res_arg, const ExprT & expr_param, const PatT & pat_param, const PosT & pos_param, const OccurT & occur_param, const MatchTypeT & match_type_param) const
+    {
+        size_t col_size = expr_param.getDataNum();
+
+        // Get function pointers to process the specific int type
+        GetIntFuncPointerType get_pos_func = getGetIntFuncPointer(pos_param.getIntType());
+        GetIntFuncPointerType get_occur_func = getGetIntFuncPointer(occur_param.getIntType());
+
+        // Container will not be used when parm is const
+        const void * pos_container = pos_param.getContainer();
+        const void * occur_container = occur_param.getContainer();
+
+        // Const value will not be used when param is not const
+        Int64 pos_const_val = PosT::isConst() ? pos_param.template getInt<Int64>(0) : -1;
+        Int64 occur_const_val = OccurT::isConst() ? occur_param.template getInt<Int64>(0) : -1;
+
+        // Check if args are all const columns
+        if constexpr (ExprT::isConst() && PatT::isConst() && PosT::isConst() && OccurT::isConst() && MatchTypeT::isConst())
+        {
+            if (expr_param.isNullAt(0) || pat_param.isNullAt(0) || pos_param.isNullAt(0) || occur_param.isNullAt(0) || match_type_param.isNullAt(0))
+            {
+                res_arg.column = res_arg.type->createColumnConst(col_size, Null());
+                return;
+            }
+
+            String pat = pat_param.getString(0);
+            if (unlikely(pat.empty()))
```

Comment:
Do we need to check whether pos >= 1?

Reply:
> Do we need to check whether pos >= 1?

Yes, we check position in this file `OptimizedRegularExpression.inl.h`.